File: tmp_number_of_islands_ok.py
from typing import List

def numIslands(grid: List[List[str]]) -> int:
    if not grid:
        return 0
    m, n = len(grid), len(grid[0])
    seen = [[False] * n for _ in range(m)]

    def dfs(sr, sc):
        stack = [(sr, sc)]
        seen[sr][sc] = True
        while stack:
            r, c = stack.pop()
            for dr, dc in ((1,0),(-1,0),(0,1),(0,-1)):
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and not seen[nr][nc] and grid[nr][nc] == "1":
                    seen[nr][nc] = True
                    stack.append((nr, nc))

    count = 0
    for i in range(m):
        for j in range(n):
            if grid[i][j] == "1" and not seen[i][j]:
                count += 1
                dfs(i, j)
    return count

File: tmp_number_of_islands_ok.java
import java.util.*;

public class Solution {
    public static int numIslands(String[][] grid) {
        if (grid == null || grid.length == 0) return 0;
        int rows = grid.length;
        int cols = grid[0].length;
        boolean[][] seen = new boolean[rows][cols];
        int count = 0;
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (!seen[r][c] && "1".equals(grid[r][c])) {
                    count++;
                    dfs(grid, seen, r, c);
                }
            }
        }
        return count;
    }

    private static void dfs(String[][] grid, boolean[][] seen, int r, int c) {
        int rows = grid.length;
        int cols = grid[0].length;
        if (r < 0 || c < 0 || r >= rows || c >= cols) return;
        if (seen[r][c] || !"1".equals(grid[r][c])) return;
        seen[r][c] = true;
        dfs(grid, seen, r + 1, c);
        dfs(grid, seen, r - 1, c);
        dfs(grid, seen, r, c + 1);
        dfs(grid, seen, r, c - 1);
    }
}
